src/libs/type-utils/lodash-fp-helpers.ts

```diff
@@ -0,0 +1,138 @@
+import { AnyFn, AnyPromiseFn, GenericFn, WrapFn } from 'src/libs/type-utils/general-types'
+
+/**
+ * (async Promise overload) Provides better type flow and ergonomics for the common case of wrapping a
+ * main function (typically a data call that returns a promise) with
+ * additional pass-thru wrapper functions (handlers) to handle things like
+ * errors or retry.  These handlers are expected to be HoF's
+ * (higher order functions) that return a wrapped function with the desired
+ * handling logic which still honors the same function signature of the fn
+ * it is given as it's single argument
+ * @param handlers
+ * @param mainFn
+ */
+export function withHandlers<P, A extends any[], F extends (...args: A) => Promise<P>>(
+  handlers: WrapFn<(...args: A) => Promise<P | unknown>>[],
+  mainFn: F
+): F
+/**
+ * (non-async overload) Provides better type flow and ergonomics for the
+ * common case of wrapping a main function with additional pass-thru wrapper
+ * functions (handlers).  These handlers are expected to be HoF's
+ * (higher order functions) that return a wrapped function with the desired
+ * handling logic which still honors the same function signature of the fn
+ * it is given as it's single argument
+ * @param handlers
+ * @param mainFn
+ */
+export function withHandlers<F extends AnyFn, F2 extends F>(
+  handlers: WrapFn<GenericFn<F2>>[],
+  mainFn: F
+): GenericFn<F>
+export function withHandlers<F extends AnyFn>(
+  handlers: ((fn: F) => F)[],
+  mainFn: F
+): F {
+  const resultFn = _.flow(...handlers)(mainFn)
+  return resultFn
+}
+
+/**
+ * produces a curried function that expect the last argument to be given
+ * before executing the original function.
+ * Provides better type flow then the more general _.curry(fn)
+ * @param fn function with 2 arguments
+ */
+export function curryLastArg<A, LAST, R>(
+  fn: (a: A, last: LAST) => R
+) : (a: A) => (last: LAST) => R;
+/**
+ * produces a curried function that expect the last argument to be given
+ * before executing the original function.
+ * Provides better type flow then the more general _.curry(fn)
+ * @param fn function with 3 arguments
+ */
+export function curryLastArg<A, B, LAST, R>(
+  fn: (a: A, b: B, last: LAST) => R
+) : (a: A, b: B) => (last: LAST) => R;
+/**
+ * produces a curried function that expect the last argument to be given
+ * before executing the original function.
+ * Provides better type flow then the more general _.curry(fn)
+ * @param fn function with 4 arguments
+ */
+export function curryLastArg<A, B, C, LAST, R>(
+  fn: (a: A, b: B, c: C, last: LAST) => R
+) : (a: A, b: B, c: C) => (last: LAST) => R;
+/**
+ * produces a curried function that expect the last argument to be given
+ * before executing the original function.
+ * Provides better type flow then the more general _.curry(fn)
+ * @param fn function with 5 arguments
+ */
+export function curryLastArg<A, B, C, D, LAST, R>(
+  fn: (a: A, b: B, c: C, d:D, last: LAST) => R
```

Comment:
Formatting nit: space between `d:` and `D`.

Reply:
fixed